Yum uhoh (#119)
IBM/ibmi-oss-docs: troubleshooting/YUM.md

## Before you go further.....

To make sure you're seeing the latest updates available for your operating system, make sure you have
the latest version of `yum` and `ibmi-repos` installed, if you are able to. If you are unable to, read on. 
This page may still help
```
/QOpenSys/pkgs/bin/yum upgrade yum ibmi-repos
```

If you are able to successfully install the `ibmi-repos` package, you can consider removing the
legacy repo definition by running:

```
mv /QOpenSys/etc/yum/repos.d/ibm.repo /QOpenSys/etc/yum/repos.d/ibm.repo.backup
```

# Checking Connectivity

The most common cause of issue with yum is related to network connectivity. Errors will state something like
```
unable to open repomd.xml
```

The IBM server supports three protocols for downloading packages:
- HTTPs (secure, default, recommended)
- HTTP (disabled by default)
- FTP (disabled by default)

HTTP or FTP can be used as a backup mechanism if you are unable to get the default configuration working, or if your network
prohibits https connections. 

You can use Python to check connectivity to the IBM RPM server. To check if you have HTTPS connectivity and have proper
TLS setup, run:

```python

> /QOpenSys/pkgs/bin/python2.7 -c "import socket; import ssl; hostname='public.dhe.ibm.com'; ssl.create_default_context().wrap_socket(socket.create_connection((hostname,443), 30), server_hostname=hostname) ; print 'success'"

Reviewer: So one thing to note here is that our Python 2 does not do certificate validation by default, using a sitecustomize.py to override Python's default behavior as documented in [PEP 476](https://peps.python.org/pep-0476/#opting-out) (stems back to OPS, long before we had ca-certificates available.). However, yum uses pycurl, which downloads through curl which _does_ enable validation by default.

I think we'd need to undo the monkeypatch:

```suggestion
/QOpenSys/pkgs/bin/python2.7 -c "import socket; import ssl; ssl._create_default_https_context = ssl._get_https_context_factory(); hostname='public.dhe.ibm.com'; ssl.create_default_context().wrap_socket(socket.create_connection((hostname,443), 30), server_hostname=hostname) ; print 'success'"
```

Reviewer: Alternatively, maybe we make a pycurl example instead, which would be directly applicable.

Reviewer: yum doesn't actually use pycurl directly, but instead uses urlgrabber. Here's a urlgrabber one-liner: `python2 -c 'import urlgrabber; urlgrabber.urlread("https://public.dhe.ibm.com/software/ibmi/products/pase"); print "success"'`

Reviewer: This _only_ works on Python 2, however since we don't ship urlgrabber or pycurl packages for Python 3.

Author: @kadler is urlgrabber shipped in the bootstrap? I presume so as a dependency of yum..

Reviewer: Yes

```

The output from this command can help you figure out next steps:
- If the output contains `Hostname and service name not provided or found`, DNS is not configured properly
- If the output contains `timed out` or `connection refused` then you cannot reach IBM's server
- If the output contains `ssl.CertificateError: hostname '______________' doesn't match 'public.dhe.ibm.com'`, a separate entity is injecting an SSL certificate
- If the output contains another CertificateError, you need to install the `ca-certificates-mozilla` package

## Checking Connectivity for alternative protocols

**To check if you have HTTPs connectivity:**

```python
/QOpenSys/pkgs/bin/python2.7 -c "import socket; socket.create_connection(('public.dhe.ibm.com', 80), 30); print 'success'"
```

**To check for HTTP connectivity**

```python
/QOpenSys/pkgs/bin/python2.7 -c "import socket; socket.create_connection(('public.dhe.ibm.com', 80), 30); print 'success'"
```

**To check for FTP connectivity:**

```python
/QOpenSys/pkgs/bin/python2.7 -c "import socket; socket.create_connection(('public.dhe.ibm.com', 21), 30); print 'success'"
```

The output from this command can help you figure out next steps:

> - If the output contains `Hostname and service name not provided or found`, DNS is not configured properly

Reviewer: We could check for this directly using [socket.getaddrinfo](https://docs.python.org/2/library/socket.html#socket.getaddrinfo), though I suppose the above is "two birds, one stone"

- If the output contains `timed out` or `connection refused` then you cannot reach IBM's server

### DNS not configured properly

If DNS is not configured properly, please work with your IBM i system administrator or networking team to resolve the problem correctly. 

As a stopgap workaroud, you can create a host table entry for the IBM server at `public.dhe.ibm.com`.

**Important Note: This IP address may change in the future. At which point, yum will start failing and you will need to create new host table entries**

From SSH or QP2TERM:
```
system "ADDTCPHTE INTNETADR('129.35.224.112') HOSTNAME((public.dhe.ibm.com))"
```

From 5250 CL:
```
ADDTCPHTE INTNETADR('129.35.224.112') HOSTNAME((public.dhe.ibm.com))
```

### Cannot reach IBM's server

Please work with your networking team to resolve the problem.

### A separate entity is enjecting an SSL certificate

Follow [these steps](https://www.seidengroup.com/2021/04/26/how-to-validate-self-signed-ssl-tls-certificates-from-ibm-i/)
to add the new certificate as needed. 

> ## Installing ca-certificates-mozilla by disabling SSL verification
> 
> If you have the `ibmi-repos` package installed:

Reviewer: ```suggestion
## Temporarily disabling SSL verification to install ca-certificates-mozilla

To resolve certificate validation errors, usually installing ca-certificate-mozilla will resolve it, but if HTTPS is broken you can't install it resulting in a Catch-22. Temporarily disabling SSL verification can resolve the Catch-22 and allow you to install ca-certificates-mozilla.

If you have the `ibmi-repos` package installed:
```

## TL;DR Steps that will fix most people that have a working DNS setup

If you have `ibmi-repos` installed
```
/QOpenSys/pkgs/bin/yum-config-manager --disable-repo=ibm
/QOpenSys/pkgs/bin/yum-config-manager --enable-repo=ibmi-base-unsecure
/QOpenSys/pkgs/bin/yum-config-manager --enable-repo=ibmi-release-unsecure
/QOpenSys/pkgs/bin/yum install ca-certificates-mozilla
/QOpenSys/pkgs/bin/yum-config-manager --disable-repo=ibmi-base-unsecure
/QOpenSys/pkgs/bin/yum-config-manager --disable-repo=ibmi-release-unsecure
/QOpenSys/pkgs/bin/yum upgrade ibmi-repos
```

If you do not have `ibmi-repos` installed 
```

> /QOpenSys/pkgs/bin/yum-config-manager --save --setopt=ibm.sslverify=0

Reviewer: Might be worth having a command to disable this repo after ibmi-repos is installed, eg

`/QOpenSys/pkgs/bin/yum-config-manager --disable-repo=ibm`

Reviewer: Nevermind, I was just looking at the individual commit. The PR overall does do this.

/QOpenSys/pkgs/bin/yum install ca-certificates-mozilla

> /QOpenSys/pkgs/bin/yum install ibmi-repos

Reviewer: This could be one command: `install ca-certificates-mozilla ibmi-repos`

Reviewer: ```suggestion
/QOpenSys/pkgs/bin/yum install ca-certificates-mozilla ibmi-repos
```
